File: backend/ahinlendor/action_encoding.py
from __future__ import annotations

import json
from pathlib import Path
from typing import Any

_ACTION_ENCODING_PATH = Path(__file__).resolve().parents[2] / "shared" / "action_encoding.json"

with _ACTION_ENCODING_PATH.open("r", encoding="utf-8") as _handle:
    _ACTION_ENCODING: dict[str, Any] = json.load(_handle)

BUY_FACEUP_RANGE = tuple(_ACTION_ENCODING["buyFaceup"])
BUY_RESERVED_RANGE = tuple(_ACTION_ENCODING["buyReserved"])
RESERVE_FACEUP_RANGE = tuple(_ACTION_ENCODING["reserveFaceup"])
RESERVE_DECK_RANGE = tuple(_ACTION_ENCODING["reserveDeck"])
TAKE3_RANGE = tuple(_ACTION_ENCODING["take3"])
TAKE2_SAME_RANGE = tuple(_ACTION_ENCODING["take2Same"])
TAKE2_RANGE = tuple(_ACTION_ENCODING["take2"])
TAKE1_RANGE = tuple(_ACTION_ENCODING["take1"])
PASS_RANGE = tuple(_ACTION_ENCODING["pass"])
RETURN_RANGE = tuple(_ACTION_ENCODING["return"])
NOBLE_RANGE = tuple(_ACTION_ENCODING["noble"])
CONTINUATION_RANGE = tuple(_ACTION_ENCODING["continuation"])

TAKE3_TRIPLETS: tuple[tuple[int, int, int], ...] = tuple(tuple(item) for item in _ACTION_ENCODING["take3Triplets"])
TAKE2_PAIRS: tuple[tuple[int, int], ...] = tuple(tuple(item) for item in _ACTION_ENCODING["take2Pairs"])


def _in_range(action_idx: int, action_range: tuple[int, int]) -> bool:
    return action_range[0] <= int(action_idx) <= action_range[1]


def is_buy_faceup_action(action_idx: int) -> bool:
    return _in_range(action_idx, BUY_FACEUP_RANGE)


def is_buy_reserved_action(action_idx: int) -> bool:
    return _in_range(action_idx, BUY_RESERVED_RANGE)


def is_reserve_faceup_action(action_idx: int) -> bool:
    return _in_range(action_idx, RESERVE_FACEUP_RANGE)


def is_reserve_deck_action(action_idx: int) -> bool:
    return _in_range(action_idx, RESERVE_DECK_RANGE)


def is_take3_action(action_idx: int) -> bool:
    return _in_range(action_idx, TAKE3_RANGE)


def is_take2_same_action(action_idx: int) -> bool:
    return _in_range(action_idx, TAKE2_SAME_RANGE)


def is_take2_action(action_idx: int) -> bool:
    return _in_range(action_idx, TAKE2_RANGE)


def is_take1_action(action_idx: int) -> bool:
    return _in_range(action_idx, TAKE1_RANGE)


def is_pass_action(action_idx: int) -> bool:
    return _in_range(action_idx, PASS_RANGE)


def is_return_action(action_idx: int) -> bool:
    return _in_range(action_idx, RETURN_RANGE)


def is_noble_action(action_idx: int) -> bool:
    return _in_range(action_idx, NOBLE_RANGE)


def is_continuation_action(action_idx: int) -> bool:
    return _in_range(action_idx, CONTINUATION_RANGE)

File: src/lib/actionEncoding.ts
import actionEncoding from '../../shared/action_encoding.json';
import type { ActionVizDTO } from '../types';

export type ActionColor = 'white' | 'blue' | 'green' | 'red' | 'black' | 'gold';

type ActionRange = readonly [number, number];

const COLOR_ORDER: readonly ActionColor[] = ['white', 'blue', 'green', 'red', 'black'];

export const ACTION_RANGES = actionEncoding as unknown as {
  buyFaceup: ActionRange;
  buyReserved: ActionRange;
  reserveFaceup: ActionRange;
  reserveDeck: ActionRange;
  take3: ActionRange;
  take2Same: ActionRange;
  take2: ActionRange;
  take1: ActionRange;
  pass: ActionRange;
  return: ActionRange;
  noble: ActionRange;
  continuation: ActionRange;
};

export const TAKE3_TRIPLETS = actionEncoding.take3Triplets as unknown as readonly (readonly [number, number, number])[];
export const TAKE2_PAIRS = actionEncoding.take2Pairs as unknown as readonly (readonly [number, number])[];

function isActionInRange(actionIdx: number, range: ActionRange): boolean {
  return range[0] <= actionIdx && actionIdx <= range[1];
}

export const isBuyFaceupAction = (actionIdx: number): boolean => isActionInRange(actionIdx, ACTION_RANGES.buyFaceup);
export const isBuyReservedAction = (actionIdx: number): boolean => isActionInRange(actionIdx, ACTION_RANGES.buyReserved);
export const isReserveFaceupAction = (actionIdx: number): boolean => isActionInRange(actionIdx, ACTION_RANGES.reserveFaceup);
export const isReserveDeckAction = (actionIdx: number): boolean => isActionInRange(actionIdx, ACTION_RANGES.reserveDeck);
export const isTake3Action = (actionIdx: number): boolean => isActionInRange(actionIdx, ACTION_RANGES.take3);
export const isTake2SameAction = (actionIdx: number): boolean => isActionInRange(actionIdx, ACTION_RANGES.take2Same);
export const isTake2Action = (actionIdx: number): boolean => isActionInRange(actionIdx, ACTION_RANGES.take2);
export const isTake1Action = (actionIdx: number): boolean => isActionInRange(actionIdx, ACTION_RANGES.take1);
export const isPassAction = (actionIdx: number): boolean => isActionInRange(actionIdx, ACTION_RANGES.pass);
export const isReturnAction = (actionIdx: number): boolean => isActionInRange(actionIdx, ACTION_RANGES.return);
export const isNobleAction = (actionIdx: number): boolean => isActionInRange(actionIdx, ACTION_RANGES.noble);
export const isContinuationAction = (actionIdx: number): boolean => isActionInRange(actionIdx, ACTION_RANGES.continuation);

export function getActionBankColors(action: Pick<ActionVizDTO, 'action_idx' | 'placement_hint'> | null | undefined): Set<ActionColor> {
  const out = new Set<ActionColor>();
  if (!action) {
    return out;
  }
  const idx = action.action_idx;
  if (isTake3Action(idx)) {
    for (const colorIdx of TAKE3_TRIPLETS[idx - ACTION_RANGES.take3[0]]) {
      out.add(COLOR_ORDER[colorIdx]);
    }
    return out;
  }
  if (isTake2SameAction(idx)) {
    out.add(COLOR_ORDER[idx - ACTION_RANGES.take2Same[0]]);
    return out;
  }
  if (isTake2Action(idx)) {
    for (const colorIdx of TAKE2_PAIRS[idx - ACTION_RANGES.take2[0]]) {
      out.add(COLOR_ORDER[colorIdx]);
    }
    return out;
  }
  if (isTake1Action(idx)) {
    out.add(COLOR_ORDER[idx - ACTION_RANGES.take1[0]]);
    return out;
  }
  if (isReturnAction(idx)) {
    out.add(COLOR_ORDER[idx - ACTION_RANGES.return[0]]);
    return out;
  }
  if (action.placement_hint.zone === 'bank_token' && action.placement_hint.color) {
    out.add(action.placement_hint.color as ActionColor);
  }
  return out;
}